### PR TITLE
[TypeDeclaration] Adds AddClosureParamTypeFromArgRector

### PR DESCRIPTION
**Pull request by TomasVotruba**
Rebase and merge of original PR https://github.com/rectorphp/rector-src/pull/6198 by @peterfox  :pray: 

Closes https://github.com/rectorphp/rector-src/pull/6198

### CODE REVIEW

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 1]
```diff
@@ -75,21 +75,19 @@ public function refactor(Node $node): ?Node
     {
         $this->hasChanged = false;
         foreach ($this->addParamTypeForFunctionLikeParamDeclarations as $addParamTypeForFunctionLikeParamDeclaration) {
-            $type = match (true) {
-                $node instanceof MethodCall => $node->var,
-                $node instanceof StaticCall => $node->class,
-                default => null,
-            };
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: In this way we loose type of `$node`, that's why the `if (! ($node->name ?? null) instanceof Identifier) {` was needed.

Instead, the `instanceof` should be used to always improve knowledge about varaible types.

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 2]
```diff
@@ -125,37 +123,18 @@ private function processFunctionLike(
...
-        } else {
-            $args = array_filter($callLike->getArgs(), static function (Arg $arg) use (
-                $addParamTypeForFunctionLikeWithinCallLikeArgFromArgDeclaration
-            ): bool {
-                if ($arg->name === null) {
-                    return false;
-                }
-
-                return $arg->name->name === $addParamTypeForFunctionLikeWithinCallLikeArgFromArgDeclaration->getCallLikePosition();
-            });
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: I've removed the `string` named args feature, as those are skipped now anyway. I'd limit this rule to work only with standard args in future as well, as area of use case is pretty narrow.

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 3]
```diff
@@ -69,11 +67,12 @@ public function getNodeTypes(): array
     }
 
     /**
-     * @param CallLike $node
+     * @param MethodCall|StaticCall $node
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: It's always better to have exact same types as in `getNodeTypes()`. The `CallLike` is to generic and it gives the `refactor()` method types that are never passed.

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 4]
```diff
@@ -115,8 +114,8 @@ public function configure(array $configuration): void
         $this->addParamTypeForFunctionLikeParamDeclarations = $configuration;
     }
 
-    private function processFunctionLike(
-        CallLike $callLike,
+    private function processCallLike(
+        MethodCall|StaticCall $callLike,
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: Result of well defined types in `refactor()` docblock :+1:

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 5]
```diff
@@ -171,16 +170,13 @@ private function refactorParameter(Param $param, Arg $arg): void
...
-        if (
-            $paramOrigin instanceof ClassConstFetch &&
-            $paramOrigin->name instanceof Identifier &&
-            $paramOrigin->class instanceof Name &&
-            $paramOrigin->name->name === 'class') {
-            $objectType = new ObjectType($paramOrigin->class->toString());
-        } else {
+        $argType = $this->nodeTypeResolver->getType($arg->value);
+        if (! $argType instanceof ConstantStringType) {
             return;
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: PHPStan can help us with class name detection :+1:

File: rules-tests/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector/Fixture/fixture.php.inc [hunk 6]
```diff
@@ -15,10 +14,9 @@ SomeClass::someCall(SomeClass::class, function ($object) {
 
 namespace Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddClosureParamTypeFromArgClassStringRector\Fixture;
 
-use SomeNamespace\SomeClass;
-use SomeNamespace\AnotherClass;
+use Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddClosureParamTypeFromArgClassStringRector\Source\SimpleContainer;
 
-SomeClass::someCall(SomeClass::class, function (\SomeNamespace\SomeClass $object) {
+SimpleContainer::someCall(\Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddClosureParamTypeFromArgClassStringRector\Source\SomeType::class, function (\Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddClosureParamTypeFromArgClassStringRector\Source\SomeType $object) {
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector/Fixture/fixture.php.inc]: By defalut, non-existing class are seen as strings. By both PHPStan and Rector.

That's why it's better to use existing class in tests, so we know it's an object and not a dummy string.

File: rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php [hunk 7]
```diff
@@ -1,201 +0,0 @@
...
-use Rector\StaticTypeMapper\StaticTypeMapper;
-use Rector\TypeDeclaration\ValueObject\AddClosureParamTypeFromArg;
-use Symplify\RuleDocGenerator\ValueObject\CodeSample\ConfiguredCodeSample;
-use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
-use Webmozart\Assert\Assert;
-
-/**
- * @see \Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddClosureParamTypeFromArgRector\AddClosureParamTypeFromArgRectorTest
- */
-final class AddClosureParamTypeFromArgClassStringRector extends AbstractRector implements ConfigurableRectorInterface
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromArgClassStringRector.php]: We can separate `string` and object type now, so both rule now do the same.  I've merged them together.

File: rector.php [hunk 8]
```diff
@@ -46,6 +48,14 @@
         '*/Source*',
         '*/Expected/*',
 
+        // stmts aware type
+        RenameParamToMatchTypeRector::class => [
+            __DIR__ . '/rules/DeadCode/Rector/If_/RemoveUnusedNonEmptyArrayBeforeForeachRector.php',
+        ],
+        AddMethodCallBasedStrictParamTypeRector::class => [
+            __DIR__ . '/rules/DeadCode/Rector/If_/RemoveUnusedNonEmptyArrayBeforeForeachRector.php',
+        ],
```
Comment (samsonasik) [inline rector.php:57]: this is not needed, I tested latest main branch without this skip and it is working ok, I will create PR to remove it.

Comment (samsonasik) [reply rector.php:57]: see https://github.com/rectorphp/rector-src/pull/6262, tested with `rm -rf vendor && composer update && bin/rector`, nothing change on rules directory

Comment (TomasVotruba) [reply rector.php:57]: I had some local glitches with PHPStan cache.
Patches get stuck sometimes:)

Thanks for restoring 👌

Comment (samsonasik) [reply rector.php:57]: You may need to force remove it

```
php -r 'shell_exec("rm -rf " . sys_get_temp_dir() . "/rector_cached_files");';
php -r 'shell_exec("rm -rf " . sys_get_temp_dir() . "/cache/PHPStan");';
php -r 'shell_exec("rm -rf " . sys_get_temp_dir() . "/cache/nette.configurator");';
```

**Comment by TomasVotruba** [comment]
I'm going through some old rules and see if they're used. Goal is to push too specific rules down to local custom or external projects, to make Rector core leaner and more universal to typical user.

I could not  find these 2 on Github, nor Rector Laravel:

* https://github.com/search?q=AddClosureParamTypeFromArgRector%3A%3Aclass&type=code
* https://github.com/search?q=AddClosureParamTypeFromArgClassStringRector%3A%3Aclass&type=pullrequests
* https://github.com/search?q=repo%3Adriftingly%2Frector-laravel%20AddClosureParamTypeFromArgRector&type=code

Saying that, I'll deprecate these 2 and remove them. Just headsup :+1: 

If they're used, I recommend adding them to Laravel specific sets. It might be easier to add these types, without having exact `Application` type. I know Laravel uses lot of magic and types can be lost easily. But this rule would be still useful in `ServiceProvider` context.

**Pull request merged**